motion_computation/src/motion_computation_worker.cpp

```diff
@@ -129,4 +138,218 @@ void MotionComputationWorker::setConfidenceDropRate(double drop_rate)
+
+  auto prev_pt_map = local_projector_->projectECEF(prev_pt_ecef, -1);
+
+  for (size_t i = 1; i < msg.trajectory.offsets.size(); i ++)
+  {
+    auto curr_pt_msg = msg.trajectory.offsets[i];
+    lanelet::BasicPoint3d curr_pt_ecef {ecef_x + (double)curr_pt_msg.offset_x /100.0, ecef_y + (double)curr_pt_msg.offset_y /100.0, ecef_z + (double)curr_pt_msg.offset_z /100.0};
+    auto curr_pt_map = local_projector_->projectECEF(curr_pt_ecef, -1);
+
+    cav_msgs::PredictedState curr_state;
+
+    if (i == 1) // First point's state should be stored outside "predictions"
+    {
+      curr_state = composePredictedState(curr_pt_map, prev_pt_map, output.header.stamp); 
+      output.pose.pose = curr_state.predicted_position;
+      output.velocity.twist = curr_state.predicted_velocity;
+    }
+    else    
+    {
+      curr_state = composePredictedState(curr_pt_map, prev_pt_map, prev_state.header.stamp + ros::Duration(mobility_path_prediction_time_step_));
+      output.predictions.push_back(curr_state);
+    }
+
+    if (i == msg.trajectory.offsets.size() - 1) // if last point, copy the prev_state velocity & orientation to the last point too
+    {
+      curr_state = composePredictedState(curr_pt_map, prev_pt_map, curr_state.header.stamp + ros::Duration(mobility_path_prediction_time_step_));
+      curr_state.predicted_position.position.x = curr_pt_map.x();
+      curr_state.predicted_position.position.y = curr_pt_map.y();
+      curr_state.predicted_position.position.z = curr_pt_map.z();
+      output.predictions.push_back(curr_state);
+    }
+    
+    prev_state = curr_state;
+    prev_pt_map = curr_pt_map;
+  }
+
+  return output;
+}
+
+cav_msgs::PredictedState MotionComputationWorker::composePredictedState(const lanelet::BasicPoint3d& curr_pt, const lanelet::BasicPoint3d& prev_pt, const ros::Time& prev_time_stamp)
+{
+  cav_msgs::PredictedState output_state;
+  // Set Position
+  output_state.predicted_position.position.x = prev_pt.x();
+  output_state.predicted_position.position.y = prev_pt.y();
+  output_state.predicted_position.position.z = prev_pt.z();
+
+  // Set Orientation
+  Eigen::Vector2d vehicle_vector = {curr_pt.x() - prev_pt.x() ,curr_pt.y() - prev_pt.y()};
+  Eigen::Vector2d x_axis = {1, 0};
+  double yaw = std::acos(vehicle_vector.dot(x_axis)/(vehicle_vector.norm() * x_axis.norm()));
+
+  tf2::Quaternion vehicle_orientation;
+  vehicle_orientation.setRPY(0, 0, yaw);
+  output_state.predicted_position.orientation.x = vehicle_orientation.getX();
+  output_state.predicted_position.orientation.y = vehicle_orientation.getY();
+  output_state.predicted_position.orientation.z = vehicle_orientation.getZ();
+  output_state.predicted_position.orientation.w = vehicle_orientation.getW();
+
+  // Set velocity
+  output_state.predicted_velocity.linear.x = vehicle_vector.norm() / mobility_path_prediction_time_step_;
```

Comment:
It might also worth setting the angular.z velocity as the rate of change of angle between this and the last point.

Reply:
I was not sure if I could put it in here because I needed next prediction's orientation to calculate the angular speed. Next prediction's orientation only gets calculated on the next iteration, so I added extra function to loop over all the predictions and fill in the angular velocity. Lmk if that makes sense

Reply:
I'm good with that.